adapters/http-base/src/test/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapterTest.java

```diff
@@ -359,6 +370,158 @@ public void testUploadEventWaitsForAcceptedOutcome() {
+        final HttpServerResponse response = mock(HttpServerResponse.class);
+        final SpanContext spanContext = mock(SpanContext.class);
+        final Span span = mock(Span.class);
+        when(span.context()).thenReturn(spanContext);
+
+        final String tenant = "tenant";
+        final String device = "device";
+        final String contentType = "application/text";
+
+        final PubSubBasedCommand command = mock(PubSubBasedCommand.class);
+        when(command.isAckRequired()).thenReturn(false);
+
+        final AbstractCommandContext<PubSubBasedCommand> abstractCommandContext = mock(AbstractCommandContext.class);
+        when(abstractCommandContext.getCommand()).thenReturn(command);
+        when(abstractCommandContext.getTracingSpan()).thenReturn(span);
+
+        final HttpContext ctx = newHttpContext(payload, contentType, newEventRequest(), response, abstractCommandContext);
+        when(ctx.get(CommandContext.KEY_COMMAND_CONTEXT)).thenReturn(abstractCommandContext);
+        when(ctx.getRoutingContext().addBodyEndHandler(VertxMockSupport.anyHandler())).thenAnswer(invocation -> {
+            final Handler<Void> handler = invocation.getArgument(0);
+            handler.handle(null);
+            return 0;
+        });
+
+        adapter.doUploadMessage(ctx, tenant, device);
+        assertEventHasBeenSentDownstream(tenant, device, contentType);
+        // THEN the device does not get a response
+        verify(response, never()).end();
+        // and the message is not reported as being processed
+        verify(metrics, never()).reportTelemetry(
+                any(MetricsTags.EndpointType.class),
+                anyString(),
+                any(),
+                eq(MetricsTags.ProcessingOutcome.FORWARDED),
+                any(MetricsTags.QoS.class),
+                anyInt(),
+                any(MetricsTags.TtdStatus.class),
+                any());
+
+        // until the event has been accepted
+        outcome.complete();
+        verify(response).end();
+        verify(metrics).reportTelemetry(
+                eq(MetricsTags.EndpointType.EVENT),
+                eq(tenant),
+                any(),
+                eq(MetricsTags.ProcessingOutcome.FORWARDED),
+                eq(MetricsTags.QoS.AT_LEAST_ONCE),
+                eq(payload.length()),
+                eq(MetricsTags.TtdStatus.NONE),
+                any());
+
+        verify(abstractCommandContext, times(2)).getCommand();
+        verify(abstractCommandContext).getTracingSpan();
+        verify(command).isAckRequired();
+    }
+
+    /**
+     * Verifies that the adapter waits for an event with an associated command (ackRequired=true) being settled and accepted
+     * by a downstream peer before responding with a 202 status to the device and sending a command response to the
+     * messaging infrastructure.
+     */
+    @Test
+    public void testUploadEventWithAckRequiredCommandWaitsForAcceptedOutcome() {
```

Comment:
I am also not sure if I understand the use case here. Can you elaborate (and revisit the test case documentation) accordingly? How is this related to the feature being added to the HTTP adapter?

Reply:
see above